app/services/places.js

```diff
@@ -0,0 +1,95 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+import {
+  datascript,
+  mori,
+  helpers,
+} from 'datascript-mori';
+import { DB } from './user-agent-service/sqlite';
+
+const { vector, parse, conj } = mori;
+const { DB_ADD } = helpers;
+
+export class PlacesImporter {
+  constructor(placesDB) {
+    this.db = placesDB;
+    this.places = new Map();
+  }
+
+  async buildPlaceMap(limit = undefined) {
+    const placeQuery = `SELECT id, url, title, visit_count, last_visit_date, guid
+    FROM moz_places
+    WHERE hidden = 0
+    ${limit ? ' LIMIT ${limit}' : ''}
+    `;
+    await this.db.each(placeQuery, [], (row) => {
+      this.places.set(row.id, {
+        id: row.id,
+        url: row.url,
+        title: row.title,
+        visit_count: row.visit_count,
+        last_visit_date: row.last_visit_date,
+        guid: row.guid,
+        visits: [],
+      });
+    });
+
+    const visitQuery = `SELECT id, from_visit, place_id, visit_date, visit_type, session
+    FROM moz_historyvisits`;
```

Comment:
`ORDER BY place_id, visit_date`? Then you can make the place lookup cheaper, and indeed just assign straight to `place.visits` by partitioning on `place_id` and not have to sort on `visit_date`.